File: Chagok/frontend/src/lib/api/detective-portal.ts
/**
 * Detective Portal API Functions
 * 003-role-based-ui Feature - US5
 */

import { apiRequest } from './client';
import type {
  DetectiveDashboardResponse,
  DetectiveCaseListResponse,
  DetectiveCaseDetailResponse,
  FieldRecordRequest,
  FieldRecordResponse,
  RecordPhotoUploadRequest,
  RecordPhotoUploadResponse,
  ReportRequest,
  ReportResponse,
  EarningsResponse,
  AcceptRejectResponse,
  FieldRecord,
  Transaction,
} from '@/types/detective-portal';

// Re-export types for convenience
export type { FieldRecord, Transaction, ReportRequest };
export type CaseDetailData = DetectiveCaseDetailResponse;
export type EarningsData = EarningsResponse;

/**
 * Get detective dashboard data
 */
export async function getDetectiveDashboard() {
  return apiRequest<DetectiveDashboardResponse>('/detective/dashboard', {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Get detective's case list
 */
export async function getDetectiveCases(params?: {
  status?: string;
  page?: number;
  limit?: number;
}) {
  const searchParams = new URLSearchParams();
  if (params?.status) searchParams.append('status', params.status);
  if (params?.page) searchParams.append('page', params.page.toString());
  if (params?.limit) searchParams.append('limit', params.limit.toString());

  const query = searchParams.toString();
  const url = query ? `/detective/cases?${query}` : '/detective/cases';

  return apiRequest<DetectiveCaseListResponse>(url, {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Get detective case detail
 */
export async function getDetectiveCaseDetail(caseId: string) {
  return apiRequest<DetectiveCaseDetailResponse>(`/detective/cases/${caseId}`, {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Accept an investigation case
 */
export async function acceptCase(caseId: string) {
  return apiRequest<AcceptRejectResponse>(`/detective/cases/${caseId}/accept`, {
    method: 'POST',
    credentials: 'include',
  });
}

/**
 * Reject an investigation case
 */
export async function rejectCase(caseId: string, reason: string) {
  return apiRequest<AcceptRejectResponse>(`/detective/cases/${caseId}/reject`, {
    method: 'POST',
    credentials: 'include',
    body: JSON.stringify({ reason }),
  });
}

/**
 * Create a field record for a case
 */
export async function createFieldRecord(
  caseId: string,
  data: FieldRecordRequest
) {
  return apiRequest<FieldRecordResponse>(`/detective/cases/${caseId}/records`, {
    method: 'POST',
    credentials: 'include',
    body: JSON.stringify(data),
  });
}

/**
 * Request presigned URL for record photo upload
 */
export async function requestRecordPhotoUpload(
  caseId: string,
  data: RecordPhotoUploadRequest
) {
  return apiRequest<RecordPhotoUploadResponse>(
    `/detective/cases/${caseId}/records/upload`,
    {
      method: 'POST',
      credentials: 'include',
      body: JSON.stringify(data),
    }
  );
}

/**
 * Submit investigation report
 */
export async function submitReport(caseId: string, data: ReportRequest) {
  return apiRequest<ReportResponse>(`/detective/cases/${caseId}/report`, {
    method: 'POST',
    credentials: 'include',
    body: JSON.stringify(data),
  });
}

/**
 * Get detective earnings
 */
export async function getDetectiveEarnings(period?: string) {
  const url = period
    ? `/detective/earnings?period=${period}`
    : '/detective/earnings';

  return apiRequest<EarningsResponse>(url, {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Upload photo to S3 using presigned URL
 */
export async function uploadPhoto(
  uploadUrl: string,
  file: File,
  onProgress?: (percent: number) => void
): Promise<{ success: boolean; url?: string; error?: string }> {
  return new Promise((resolve) => {
    const xhr = new XMLHttpRequest();

    xhr.upload.addEventListener('progress', (event) => {
      if (event.lengthComputable && onProgress) {
        const percent = Math.round((event.loaded / event.total) * 100);
        onProgress(percent);
      }
    });

    xhr.addEventListener('load', () => {
      if (xhr.status >= 200 && xhr.status < 300) {
        // Extract the URL without query params
        const url = uploadUrl.split('?')[0];
        resolve({ success: true, url });
      } else {
        resolve({ success: false, error: `Upload failed: ${xhr.status}` });
      }
    });

    xhr.addEventListener('error', () => {
      resolve({ success: false, error: 'Network error during upload' });
    });

    xhr.open('PUT', uploadUrl);
    xhr.setRequestHeader('Content-Type', file.type);
    xhr.send(file);
  });
}

File: Chagok/backend/app/api/detective_portal.py
"""
Detective Portal API Router
003-role-based-ui Feature - US5 (T085-T093)

API endpoints for detective portal including dashboard, case management,
field records, reports, and earnings.
"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.dependencies import get_db, require_role, get_file_storage_service
from app.core.error_messages import ErrorMessages
from app.services.detective_portal_service import DetectivePortalService
from app.services.audit_service import AuditLogService
from app.domain.ports.file_storage_port import FileStoragePort
from app.schemas.detective_portal import (
    DetectiveDashboardResponse,
    CaseListResponse,
    CaseDetailResponse,
    AcceptRejectResponse,
    RejectRequest,
    FieldRecordRequest,
    CreateRecordResponse,
    RecordPhotoUploadRequest,
    RecordPhotoUploadResponse,
    ReportRequest,
    ReportResponse,
    EarningsResponse,
    EarningsSummary,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/detective", tags=["detective-portal"])


# ============== T086: Dashboard Endpoint ==============

@router.get("/dashboard", response_model=DetectiveDashboardResponse)
async def get_detective_dashboard(
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Get detective dashboard data.

    Returns:
        - User name
        - Dashboard stats (active/pending/completed/earnings)
        - Active investigations list
        - Today's schedule
    """
    service = DetectivePortalService(db)
    try:
        return service.get_dashboard(user_id)
    except ValueError as e:
        logger.warning(f"Dashboard not found for user {user_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.DETECTIVE_DASHBOARD_NOT_FOUND
        )


# ============== T087: Case List Endpoint ==============

@router.get("/cases", response_model=CaseListResponse)
async def get_detective_cases(
    status: Optional[str] = None,
    page: int = 1,
    limit: int = 20,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Get list of detective's cases with optional filtering.

    Args:
        status: Filter by investigation status (pending/active/review/completed)
        page: Page number (default 1)
        limit: Items per page (default 20)

    Returns:
        List of cases assigned to the detective
    """
    service = DetectivePortalService(db)
    return service.get_cases(user_id, status=status, page=page, limit=limit)


# ============== T088: Case Detail Endpoint ==============

@router.get("/cases/{case_id}", response_model=CaseDetailResponse)
async def get_detective_case_detail(
    case_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Get detailed case information for detective.

    Args:
        case_id: Case ID to retrieve

    Returns:
        Detailed case information including records
    """
    service = DetectivePortalService(db, file_storage_port=file_storage_port)
    try:
        return service.get_case_detail(user_id, case_id)
    except PermissionError:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to this case"
        )
    except KeyError as e:
        logger.warning(f"Case not found: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )


# ============== T089: Accept/Reject Endpoints ==============

@router.post("/cases/{case_id}/accept", response_model=AcceptRejectResponse)
async def accept_case(
    case_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Accept an investigation case.

    Args:
        case_id: Case ID to accept

    Returns:
        Success response with new status
    """
    service = DetectivePortalService(db)
    audit_service = AuditLogService(db)

    try:
        result = service.accept_case(user_id, case_id)

        # T091: Audit logging
        audit_service.log_action(
            user_id=user_id,
            action="case_accept",
            resource_type="case",
            resource_id=case_id,
            details={"action": "accept"}
        )

        return result
    except KeyError as e:
        logger.warning(f"Case accept failed: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )


@router.post("/cases/{case_id}/reject", response_model=AcceptRejectResponse)
async def reject_case(
    case_id: str,
    request: RejectRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Reject an investigation case.

    Args:
        case_id: Case ID to reject
        request: Rejection reason

    Returns:
        Success response with reason
    """
    service = DetectivePortalService(db)
    audit_service = AuditLogService(db)

    try:
        result = service.reject_case(user_id, case_id, request.reason)

        # T091: Audit logging
        audit_service.log_action(
            user_id=user_id,
            action="case_reject",
            resource_type="case",
            resource_id=case_id,
            details={"reason": request.reason}
        )

        return result
    except KeyError as e:
        logger.warning(f"Case reject failed: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )


# ============== T090: Field Record Endpoint ==============

@router.post("/cases/{case_id}/records/upload", response_model=RecordPhotoUploadResponse)
async def request_record_photo_upload(
    case_id: str,
    request: RecordPhotoUploadRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Request presigned URL for record photo upload.

    Args:
        case_id: Case ID to upload photo for
        request: Photo upload metadata

    Returns:
        Presigned S3 upload URL and object key
    """
    service = DetectivePortalService(db, file_storage_port=file_storage_port)

    try:
        return service.request_record_photo_upload(
            detective_id=user_id,
            case_id=case_id,
            file_name=request.file_name,
            content_type=request.content_type,
            file_size=request.file_size
        )
    except KeyError as e:
        logger.warning(f"Record photo upload request failed: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )
    except ValueError as e:
        logger.warning(f"Record photo upload request failed validation: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid upload request"
        )


@router.post("/cases/{case_id}/records", response_model=CreateRecordResponse)
async def create_field_record(
    case_id: str,
    request: FieldRecordRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Create a new field record for a case.

    Args:
        case_id: Case ID to add record to
        request: Record data (type, content, GPS, photo)

    Returns:
        Created record ID and success message
    """
    service = DetectivePortalService(db)
    audit_service = AuditLogService(db)

    try:
        result = service.create_field_record(
            detective_id=user_id,
            case_id=case_id,
            record_type=request.record_type,
            content=request.content,
            gps_lat=request.gps_lat,
            gps_lng=request.gps_lng,
            photo_url=request.photo_url,
            photo_key=request.photo_key
        )

        # T091: Audit logging
        audit_service.log_action(
            user_id=user_id,
            action="field_record_create",
            resource_type="investigation_record",
            resource_id=result.record_id,
            details={
                "case_id": case_id,
                "record_type": request.record_type.value,
                "has_gps": request.gps_lat is not None,
                "has_photo": request.photo_url is not None
            }
        )

        return result
    except KeyError as e:
        logger.warning(f"Field record create failed: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )
    except ValueError as e:
        logger.warning(f"Field record create failed validation: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid field record data"
        )


# ============== T092: Report Submission Endpoint ==============

@router.post("/cases/{case_id}/report", response_model=ReportResponse)
async def submit_report(
    case_id: str,
    request: ReportRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Submit final investigation report for a case.

    Args:
        case_id: Case ID to submit report for
        request: Report data (summary, findings, conclusion, attachments)

    Returns:
        Report ID and success message
    """
    service = DetectivePortalService(db)
    audit_service = AuditLogService(db)

    try:
        result = service.submit_report(
            detective_id=user_id,
            case_id=case_id,
            summary=request.summary,
            findings=request.findings,
            conclusion=request.conclusion,
            attachments=request.attachments
        )

        # T091: Audit logging
        audit_service.log_action(
            user_id=user_id,
            action="report_submit",
            resource_type="investigation_report",
            resource_id=result.report_id,
            details={
                "case_id": case_id,
                "has_attachments": bool(request.attachments)
            }
        )

        return result
    except KeyError as e:
        logger.warning(f"Report submit failed: {case_id}, error: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )


# ============== T093: Earnings Endpoint ==============

@router.get("/earnings", response_model=EarningsResponse)
async def get_detective_earnings(
    period: Optional[str] = None,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Get detective's earnings and transaction history.

    Args:
        period: Optional filter (this_month, last_month, this_year)

    Returns:
        Earnings summary and transaction list
    """
    service = DetectivePortalService(db)
    return service.get_earnings(user_id, period=period)


@router.get("/earnings/summary", response_model=EarningsSummary)
async def get_detective_earnings_summary(
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["detective"])),
):
    """
    Get detective's earnings summary (lightweight endpoint).

    Returns:
        - total_earned: Total earnings across all cases
        - pending_payment: Pending (unpaid) earnings
        - this_month: Earnings from current month
    """
    service = DetectivePortalService(db)
    return service.get_earnings_summary(user_id)
